src/python/pants/engine/legacy/graph.py

```diff
@@ -386,10 +387,11 @@ def hydrate_sources(sources_field, glob_match_error_behavior):
   # with the global default!
   path_globs = sources_field.path_globs.with_match_error_behavior(glob_match_error_behavior)
   snapshot = yield Get(Snapshot, PathGlobs, path_globs)
-  fileset_with_spec = _eager_fileset_with_spec(
+  fileset_with_spec = eager_fileset_with_spec(
     sources_field.address.spec_path,
     sources_field.filespecs,
     snapshot)
+  sources_field.validate_fn(fileset_with_spec)
```

Comment:
If you have appetite for it you could lean more into the design of #4535, and do subclassing of `Field` to allow for additional field types and validations. Someone who wanted to validate "singular" vs "plural" could subclass `Field` to do it. A sticking point for that is allowing for installation of the additional field types so that plugins can do subclassing of `Field`: see the last comment on https://github.com/pantsbuild/pants/issues/4535#issuecomment-380654441 for more context.

Reply:
I like the idea, but I'm going to punt on this for now :)